oak/server/channel.h

```diff
@@ -23,123 +23,157 @@
-// non-zero and indicates the required size for the message.  Otherwise,
-// |required_size| will be zero and |data| holds the message (transferring
-// ownership).
+// bigger than the requested maximum size, then |required_size| will be non-zero
+// and indicates the required size for the message.  If the operation would have
+// been accompanied by more than the requested maximum channel count, then
+// |required_channels| will be non-zero and indicates the required channel count
+// for the message. Otherwise, |required_size| and |required_channels| will be
+// zero and |data| holds the message (transferring ownership).
 struct ReadResult {
   uint32_t required_size;
-  std::unique_ptr<Message> data;
+  uint32_t required_channels;
+  std::unique_ptr<Message> msg;
 };
 
-// Abstract interface for Oak communication channels.
-//
-// A Channel represents a uni-directional stream of messages. Each Channel has
-// two halves, a send half and a receive half.  A channel half is represented by
-// a common ChannelHalf type so that they can be referenced from a common
-// numbering space, even though the send and receive halves have distinct
-// functionality.
+class MessageChannelWriteHalf;  // forward declaration
+class MessageChannelReadHalf;   // forward declaration
+
+// A channel, which holds an arbitrary number of queued messages in a
+// uni-directional stream.
 //
 // No flow control is implemented at this level; application
 // may decide to build some of these abstractions on top of the Channel
 // interface.
 //
 // Each channel may be connected to a built-in component, or to a local or
 // remote Oak Node.
+// (in future) remote Oak Node.
 //
 // TODO: add a hard limit for message size
-class ChannelHalf {
+class MessageChannel {
  public:
-  virtual ~ChannelHalf() {}
-
-  // Read a message of up to |size| bytes from the Channel. The caller owns any
-  // returned message, whose actual size of may be less than |size|.  If the
-  // next available message on the channel is larger than |size|, no data will
-  // be returned and the required_size field of the result will indicate the
-  // required size.
-  virtual ReadResult Read(uint32_t size) {
-    // Fallback implementation that returns an empty message; attempting to
-    // read from the send half of a channel will reach this fallback.
-    ReadResult nothing;
-    nothing.required_size = 0;
-    return nothing;
-  }
+  struct ChannelHalves {
+    std::unique_ptr<MessageChannelWriteHalf> write;
+    std::unique_ptr<MessageChannelReadHalf> read;
+  };
 
-  // BlockingRead behaves like Read but blocks until a message is available.
-  virtual ReadResult BlockingRead(uint32_t size) { return Read(size); }
+  // Create a message channel and return references to both halves of it.
+  static ChannelHalves Create();
 
-  // Indicate whether a Read operation would return a message.
-  virtual bool CanRead() { return false; }
+ private:
+  friend class MessageChannelReadHalf;
```

Comment:
This is pretty much saying that only these halves can use the Message. Is that correct?